heaviside/ast.py

```diff
@@ -507,3 +541,36 @@ def resolve_arns(branch, translate = lambda x, y: y):
             for branch in state.branches:
                 resolve_arns(branch, translate)
 
+def verify_goto_targets(branch):
+    """Recursivly checks that all Goto states target valid state names
+
+    Valid state names are those states in the current branch. This means that
+    a Goto cannot target states in another parallel branch or from a parallel
+    branch to the main body of the Step Function
+
+    Args:
+        branch (list): List of ASTState objects
+
+    Raises:
+        CompileError : If a Goto state targets an invalid state
+    """
+    if not hasattr(branch, 'states'):
+        branch.raise_error("Trying to check goto targets for non-branch state")
+
+    to_process = [branch.states]
+
+    while len(to_process) > 0:
+        states = to_process.pop(0)
+
+        names = set() # Need to know all of the valid state names for the branch
+        for state in states:
+            names.add(state.name)
```

Comment:
Are we checking somewhere else that this branch of states doesn't contain two states with the same name?  I don't see that check in this function?

Reply:
Line 481 `def check_names(branch)` is the check that verifies state names. It is run before this check, but could be run afterwards. I wanted to keep the check functions simple and didn't want to co-mingle the logic of the two checks.

Reply:
Cool.